### DISCUSSION TITLE
substituting real values for % terms in an html document on a Pico webpage server

### DISCUSSION THREAD

**Discussion by Buddles52** [micropython/micropython#16251] [Other]
This is my html code
# create the html document template
html = """<!DOCTYPE html><html>
<head><meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="icon" href="data:,">
<style>html { font-family: Helvetica; display: inline-block; margin: 0px auto; text-align: center;}
.buttonGreen { background-color: #4CAF50; border: 2px solid #000000;; color: white; padding: 15px 32px; text-align: center; text-decoration: none; display: inline-block; font-size: 16px; margin: 4px 2px; cursor: pointer; }
.buttonRed { background-color: #D11D53; border: 2px solid #000000;; color: white; padding: 15px 32px; text-align: center; text-decoration: none; display: inline-block; font-size: 16px; margin: 4px 2px; cursor: pointer; }
text-decoration: none; font-size: 30px; margin: 2px; cursor: pointer;}
</style></head>
<body><center><h1>Control Panel</h1></center><br><br>
<form><center>
<center> <button class="buttonGreen" name="LED" value="on" type="submit">LED ON</button>
<br><br>
<center> <button class="buttonRed" name="LED" value="off" type="submit">LED OFF</button>
</form>
<br><br>
<br><br>
<p>%s1<p>
<p>%s2<p>
"""
This is my Tonny micropython server code
s1 = LEDState\
                  + " and " + buttonState\
                  + " and internal temp is " + PicoTemp\
                  + " with distance " + str(distance)\
                  + " dated "\
                  +  str(MyTime[0:3])
        s2 = "second element"
        s1 = LEDState + " and " + buttonState
        response = html % s1 % s2   ********************************************this line throws the syntax error
        cl.send('HTTP/1.0 200 OK\r\nContent-type: text/html\r\n\r\n')
        cl.send(response)
        
How do I get new values for both s1 and s2 into the html doc ??  help please

**Comment by Buddles52** [comment]
Profound Apologies for the very poor formatting of my question here.  I will do better next time

**Reply by GitHubsSilverBullet** [reply]
You can do now, the edit option is available.

**Reply by Buddles52** [reply]
Sorry, my question isnt clear.  I am using the Pico as a webserver and it should send an html document to a client on request.  At the time of sending, the html document variables s1 and s2 are updated by the Pico  so that the client always receives the current values of s1 and s2.  Using the % symbol in my micropython code (Thonny) allows me to substitute a new value for s1 in the html document perfectly well.  But I want to update two values s1 AND s2 in the html deocument.  At this point i get a syntax error for the line marked by ***** above.  
So - how do I replace multiple values in an html document stored on a Pico in real time when the doc is requested by a client?

**Comment by GitHubsSilverBullet** [comment]
Should you expect more help, I'd **strongly** suggest to read about formatting your code properly.
It's all up there: **[Welcome (Please read before posting)](https://github.com/orgs/micropython/discussions/9111)**

But here's how to replace values.
```py
html_template = 'start     %s1     %s2     end'
print(html_template.replace('%s1', 'XXX').replace('%s2', 'YYY'))
```

or (even easier):
```py
html_template = 'start     {}     {}     end'
print(html_template.format('XXX', 'YYY'))
```

**Comment by Buddles52** [comment]
Hi
Thank you very much indeed for your response.  I Take your point about formatting.
I will use your code asap.  I know I can easily find out by trying it but just for the record, would I be right in thinking I can replace s1 and s2 with variables instead of text strings?

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by sosi-deadeye** [comment]
You can use any kind of object. If an object has a `__str__` method, the output of this method is used to fill in the text. If the object has not a `__str__` method, you'll get the representation of an object (`__repr__`).

The `%` format method should not be used for formatting. It's still available for compatibility reasons.
Use instead format strings to substitute strings.

Or you could use https://github.com/pfalcon/utemplate

**Comment by Buddles52** [comment]
Hi
Thank you very much indeed for responding.  I’ll try these out and hope to have some feedback soon.

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________
